### ISSUE TITLE
ENH: better customisation of a categorical legend in explore()

### ISSUE THREAD

**Issue by martinfleis** [geopandas/geopandas#2330]
We should allow some form of customisation of the categorical legend in `explore()` as was pointed out in this [SO](https://stackoverflow.com/questions/70887834/geopandas-explore-reorder-items-in-legend/70892536#70892536) question.

At the moment, we don't even adhere to `labels` passed in the `legend_kwds`. We may even think about exposing `_categorical_legend` as a public API.

**Comment by rraymondgh** [comment]
I have been looking into this.  Start by setting up a correct categorical legend plot:

```
import geopandas as gpd
import numpy as np
import pandas as pd
from geopandas.explore import _categorical_legend

world = gpd.read_file(gpd.datasets.get_path("naturalearth_lowres"))

world = world.loc[
    world["continent"].eq("Europe") & ~world["iso_a3"].isin(["RUS", "-99"])
]

colors = [
    "#3366CC",
    "#DC3912",
    "#FF9900",
    "#109618",
    "#990099",
    "#0099C6",
    "#DD4477",
    "#66AA00",
    "#B82E2E",
    "#316395",
]
bins = pd.qcut(world["pop_est"], q=5, retbins=True)[1].tolist()
bins = [0] + [round(b, -5) for b in bins[1:]]
labels = [
    f"{str(a/10**6)+'m - ' if a>0 and b!=bins[-1] else 'Greater than ' if b==bins[-1] else 'Less than '}{b/10**6 if b<bins[-1] else a/10**6}m"
    for a, b in zip(bins, bins[1:])
]
world["pop"] = pd.cut(world["pop_est"], bins=bins, labels=labels)

# correct as categorical is being used appropriately
m = world.explore(column="pop", legend=True, cmap=colors[:5], height=300, width=500)
m
```
<img width="511" alt="image" src="https://user-images.githubusercontent.com/42769112/158172355-365b5855-5e2f-4dfb-95fe-f1d5072c638c.png">

Bash categorical column so it is a string.  Results in alphanumeric sort order
```
world["pop string"] = pd.cut(world["pop_est"], bins=bins, labels=labels).astype(str)

# order of legend is effectively alphanumeric sort order
m = world.explore(
    column="pop string", legend=True, cmap=colors[:5], height=300, width=500
)

m
```
<img width="508" alt="image" src="https://user-images.githubusercontent.com/42769112/158172540-cdc92d1a-de0d-4c7d-b01d-5d158b7ff446.png">

Now how to go from string based categories to defined sort order.  IMHO best to use `pd.Categorical()`
```
# order categories with defined order - let's reverse the order
world["reverse"] = pd.Categorical(world["pop string"], categories=labels[::-1], ordered=True)

m = world.explore(
    column="reverse", legend=True, cmap=colors[:5], height=300, width=500
)

m
```
<img width="512" alt="image" src="https://user-images.githubusercontent.com/42769112/158172829-4994fd07-fde2-4d12-ba99-4c20493b05da.png">

Now this leads a question - is there any point in providing capability at API of `explore()`?  It could be just a documentation item, want a *custom* sort then caller should create a **pandas categorical** column and use that.

**Comment by rraymondgh** [comment]
In going through this exercise (plus other times I have used `explore()`) a different observation, when using **height** and **width** parameters the legend is not bound to the extent of the **folium** `div`.  I think this is a valid patch:

```
@@ -717,6 +717,10 @@ def _categorical_legend(m, title, categories, colors):
             }
         });
     });
+    // re-parent legend into folium map so height and width positioning works
+    $( document ).ready(function() {
+        $(".maplegend").detach().appendTo(".folium-map");
+    });
     </script>
     <style type='text/css'>
       .maplegend {
```

It re-parents the legend to be contained within the `<div>`  created by **folium**.  Thoughts?   (This can be seen in all screen shots I have included above) Happy to generate PR
